### ISSUE TITLE
Metrics gradually disappearing over time

### ISSUE THREAD

**Issue by aobertyshev** [ivx/yet-another-cloudwatch-exporter#401]
After running for 10 minutes (and thus doing 2 requests to cloudwatch), the exporter slowly starts reporting NaN as the values of the metrics.
![image](https://user-images.githubusercontent.com/4324562/121687744-fbafca00-cac2-11eb-8a48-887bf92b3c75.png)

As you can see, at 01:41 there are still some values left, but 5 minutes later even they are missing.
This is what the metrics page of the exporter looks like after this:
![image](https://user-images.githubusercontent.com/4324562/121688078-55b08f80-cac3-11eb-88a8-d134d7b31a59.png)

I found this comment https://github.com/ivx/yet-another-cloudwatch-exporter/issues/378#issuecomment-843487412
But after setting `period` to 330 and restarting the service, my aws_ec2_cpucredit_usage_average metrics simply disappeared.

Here's my config:
`
---
discovery:
  jobs:
  - regions:
      - eu-central-1
    type: ec2
    enableMetricData: true
    metrics:
      - name: CPUCreditUsage
        statistics:
        - Average
        period: 300
        length: 600
`
And my systemd service file:
`
[Service]
ExecStart=/usr/local/bin/yace --config.file=/etc/cloudwatch_exporter/example.yml
Environment="AWS_ACCESS_KEY_ID=<secret>"
Environment="AWS_SECRET_ACCESS_KEY=<secret>"

[Install]
WantedBy=multi-user.target
`

I would appreciate any insight.

**Comment by abhishek-annigeri** [comment]
@aobertyshev Do you see any warnings or error messages in YACE logs?

**Comment by aobertyshev** [comment]
No, here's the log output (`journalctl -u cloudwatch_exporter`) from restarting it just now:
```
-- Logs begin at Mon 2021-06-14 10:47:15 UTC, end at Wed 2021-06-16 06:49:34 UTC. --
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Stopping cloudwatch_exporter.service...
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Stopped cloudwatch_exporter.service.
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Started cloudwatch_exporter.service.
Jun 16 06:49:34 aws-frt-prometheus yace[10670]: {"level":"info","msg":"Parse config..","time":"2021-06-16T06:49:34Z"}
Jun 16 06:49:34 aws-frt-prometheus yace[10670]: {"level":"info","msg":"Startup completed","time":"2021-06-16T06:49:34Z"}
```

The metrics appear:
![image](https://user-images.githubusercontent.com/4324562/122171368-ee5a5d00-ce7f-11eb-8eb0-c659ac0a0f20.png)

Here's the /metrics page:
![image](https://user-images.githubusercontent.com/4324562/122172323-efd85500-ce80-11eb-8dd1-6d24ab8a8661.png)

2 `NaN` here are from the VMs that are not running.

5 min later:
Logs:
```
-- Logs begin at Mon 2021-06-14 10:47:15 UTC, end at Wed 2021-06-16 06:55:09 UTC. --
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Stopping cloudwatch_exporter.service...
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Stopped cloudwatch_exporter.service.
Jun 16 06:49:34 aws-frt-prometheus systemd[1]: Started cloudwatch_exporter.service.
Jun 16 06:49:34 aws-frt-prometheus yace[10670]: {"level":"info","msg":"Parse config..","time":"2021-06-16T06:49:34Z"}
Jun 16 06:49:34 aws-frt-prometheus yace[10670]: {"level":"info","msg":"Startup completed","time":"2021-06-16T06:49:34Z"}
```

/metrics page:
![image](https://user-images.githubusercontent.com/4324562/122172634-3d54c200-ce81-11eb-987a-448cfd13f41a.png)

You can see some `NaN` appearing.

5 more mins later (thus 10 minutes after restarting):
Log stayed the same.

/metrics page:
![image](https://user-images.githubusercontent.com/4324562/122172975-93296a00-ce81-11eb-87bd-c5226c079cb5.png)

Grafana:
![image](https://user-images.githubusercontent.com/4324562/122173073-ab998480-ce81-11eb-80fd-2d37f8a11207.png)

Some additional info:

- I'm running this version https://github.com/ivx/yet-another-cloudwatch-exporter/releases/tag/v0.27.0-alpha
- System info (`uname -a`): `Linux <hostname> 4.9.0-14-amd64 #1 SMP Debian 4.9.246-2 (2020-12-17) x86_64 GNU/Linux` (Debian 9)
- If I restart the exporter, I receive the metrics again
- IAM permissions of the account that is being used by the exporter: 
![image](https://user-images.githubusercontent.com/4324562/122172177-cc150f00-ce80-11eb-974d-c9715507a7a5.png)

**Comment by abhishek-annigeri** [comment]
Even I get the same thing when I add new resources. Until I restart the exporter, the YACE won't get the data.

**Comment by marekhanzlik** [comment]
Did you found any solution for this?
Only way that i was able to see stable results was with setting `length: 300` `period: 600`
but then `cloudwatch-exporter` logged warning 
```
length(300) is smaller than period(600). This can cause that the data requested is not ready and generate data gaps
```

**Comment by aobertyshev** [comment]
My temporary solution is to use Cloudwatch plugin for Grafana https://grafana.com/grafana/plugins/cloudwatch/
It ships by default and you can add the Cloudwatch data source directly.

**Comment by boazreicher** [comment]
The `period` variable isn't really relevant here.  The problem is that `length` is greater than the scraping interval.
The scraping interval defaults to 300.  The code is supposed to make sure that it's equal or greater to the maximum job length, but I believe that doesn't happen due to a bug (see https://github.com/ivx/yet-another-cloudwatch-exporter/pull/406)

To fix this, you should either set the value of `length` to 300, or increase the scraping interval (`scraping-interval`) to be the same as your max length setting.
